heapster-saw/doc/tutorial/tutorial.md

```diff
@@ -0,0 +1,623 @@
+environment `env`, `is_elem` takes a 64-bit integer the "element".
+
+```LLVM
+i64 @is_elem(i64 %0, %struct.list64_t* %1)
+```
+
+Hence, precondition to the function should reflect these two inputs
+like so `arg0:int64, arg1: list int64`, for some predicates `int64`
+and `list` that express a single integer and a list of them.
+
+Moreover, the function returns a new i64-bit integer and, as we can
+see in the code of `is_elem`, the arguments are not modified. So the
+postcondition, of the function could be described by `arg0:int64,
+arg1: list int64, ret:int64`. Alternatively, if we don't care about
+the inputs after the call we can simplify it to `arg0:true, arg1:
+true, ret:int64`, where `true` is a predicate the is always satisfied.
+ 
+Then, using the linear implication `-o` we can express the
+specification of `is_elem` as
+
+```
+arg0:int64, arg1: list int64 -o arg0:true, arg1:true, ret:int64
+```
+
+Notice that this is a memory-safety specification, not a correctness
+predicate. We can reason about the correctness of the function once we
+extract it as a functional program in Coq.
+
+In the next sections, we shall define the predicates necessary to
+write this specification in Heapster.
+
+##### Defining permission predicates
+
+One of the simplest Heapster commands is `heapster_define_perm`, which defines
+a new named permission which can then be used in Heapster types. As an
+example, the following defines the permission which describes a 64-bit integer
+value:
+
+```
+heapster_define_perm env "int64" " " "llvmptr 64" "exists x:bv 64.eq(llvmword(x))"
+```
+
+The first argument is the Heapster environment, the second is its
+name, the third is its arguments (of which there are none), the fourth
+is the type of value that the permission applies to, and the fifth is
+its definition. the new permission is created and added to the
+environment.
+
+Unfortunately, there is currently no way to print the newly defined
+permissions. If you try to print the environment (`print env`) at this
+point, you will only see the `llvm` definitions. We might add
+functionality for showing permissions in the future.
+
+The permission predicate for lists is a little more complicated
+because it requires a recursive definition. To define
+[permissions](doc/Permissions.md) which can describe unbounded data
+structures, you can use the `heapster_define_recursive_perm`
+command. As an example, here is how to describe a linked list of
+64-bit words using this command:
+
+```
+heapster_define_recursive_perm
+  env
+  "List64"
+  "rw:rwmodality"
+  "llvmptr 64"
+  ["eq(llvmword(0))", "ptr((rw,0) |-> int64<>) * ptr((rw,8) |-> List64<rw>)"]
+  "List (Vec 64 Bool)"
+  "foldList (Vec 64 Bool)"
+  "unfoldList (Vec 64 Bool)";
+```
+
+Its first four arguments are the same as for `heapster_define_perm`,
+its fifth argument contains its different inductive cases (in this
+case, a `List64` is either a null pointer, or a pointer to an `Int64`
```

Comment:
Expand the part about the cases into its own 1-2 sentences describing the two different cases and what their syntax means, even if briefly. For instance, say something like "In this case, a `List64` is either a null pointer, which is expressed with the type `eq(llvmword(0))` that declares that it equal to the word value 0, or it is a pointer where offset 0 points to a value satisfying the `int64<>` type and where offset 8 points to a value satisfying the `List64<rw>` permission, i.e., it is recursively a `List64<rw>`.

Reply:
Also reiterate that the syntax `List64<rw>` is the result of applying the `List64` named permission to the single argument `rw`.